### PR TITLE
New implementation of mesh element classes

### PR DESCRIPTION
**Pull request by norihiro-w**
This PR is based on discussion in #506 and introduces new implementations of mesh element classes by combining TemplateElement class and element-specific topological/geometrical rules. 

Supporing quadratic order elements is in progress.

### CODE REVIEW

File: MeshLib/Elements/Face.h [hunk 1]
```diff
@@ -35,10 +35,10 @@ class Face : public Element
 	static const unsigned dimension = 2;
 
 	/// Get the area of this 2d element.
-	virtual double getArea() const { return _area; }
+	virtual double getArea() const { return _content; }
 
 	/// Returns the length, area or volume of a 1D, 2D or 3D element
-	double getContent() const { return _area; }
+	double getContent() const { return _content; }
```
Comment (norihiro-w) [inline MeshLib/Elements/Face.h]: getContent() is already defined in Element

File: MeshLib/Elements/FaceRule.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace MeshLib {
+
+/**
+ */
+class FaceRule
+{
+public:
+	/// Returns the face i of the element.
+	static const Element* getFace(const Element* e, unsigned i) { return e->getEdge(i); }
```
Comment (norihiro-w) [inline MeshLib/Elements/FaceRule.h]: functions here also exist in Face class

File: MeshLib/Elements/HexRule8.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	/// Returns the i-th face of the element.
+	static const Element* getFace(const Element* e, unsigned i);
+
+	/**
+	 * Checks if a point is inside the element.
+	 * @param pnt a 3D GeoLib::Point object
+	 * @param eps tolerance for numerical algorithm used or computing the property
+	 * @return true if the point is not outside the element, false otherwise
+	 */
+	static bool isPntInElement(Node const* const* _nodes, MathLib::Point3d const& pnt, double eps = std::numeric_limits<double>::epsilon());
```
Comment (norihiro-w) [inline MeshLib/Elements/HexRule8.h]: remove default parameter here and have it only in TemplateElement class

File: MeshLib/Elements/PrismRule6.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	static const unsigned n_faces = 5;
+
+	/// Constant: The number of edges
+	static const unsigned n_edges = 9;
+
+	/// Constant: The number of neighbors
+	static const unsigned n_neighbors = 5;
+
+	/// Constant: Local node index table for faces
+	static const unsigned _face_nodes[5][4];
```
Comment (norihiro-w) [inline MeshLib/Elements/PrismRule6.h]: remove prefix `_`

File: MeshLib/Elements/PyramidRule5.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		for (unsigned j=0; j<nFaceNodes; j++)
+			nodes[j] = const_cast<Node*>(e->getNode(_face_nodes[i][j]));
+
+		if (i<4)
+			return new Tri(nodes);
+		else
+			return new Quad(nodes);
+	}
+	ERR("Error in MeshLib::Element::getFace() - Index %d does not exist.", i);
+	return NULL;
```
Comment (norihiro-w) [inline MeshLib/Elements/PyramidRule5.cpp]: nullptr

File: MeshLib/Elements/PyramidRule5.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	{3, 4}  // Edge 7
+};
+
+const unsigned PyramidRule5::_n_face_nodes[5] = { 3, 3, 3, 3, 4 };
+
+unsigned PyramidRule5::getNFaceNodes(unsigned i)
+{
+	if (i<5)
+		return _n_face_nodes[i];
+	ERR("Error in MeshLib::Element::getNFaceNodes() - Index %d does not exist.", i);
```
Comment (norihiro-w) [inline MeshLib/Elements/PyramidRule5.cpp]: assert() instead of ERR?

File: MeshLib/Elements/TemplateElement.h [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+namespace MeshLib
+{
+
+/**
+ * Template for implementing mesh element classes
+ *
+ * \tparam T_BASE         Base element class, e.g. Face, Cell
+ * \tparam ELEMENT_RULE   Geometrical and topological rules of the element
+ * \tparam EDGE_RULE      Edge rule of the element
```
Comment (norihiro-w) [inline MeshLib/Elements/TemplateElement.h]: not used

File: MeshLib/Elements/TemplateElement.h [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	TemplateElement(std::array<Node*, n_all_nodes> const& nodes, unsigned value = 0, std::size_t id = std::numeric_limits<std::size_t>::max());
+
+	/// Copy constructor
+	TemplateElement(const TemplateElement &e);
+
+	/// Destructor
+	virtual ~TemplateElement() {}
+
+	/**
+	 * This method is pure virtual and is inherited from class @sa Element.
```
Comment (norihiro-w) [inline MeshLib/Elements/TemplateElement.h]: not pure virtual anymore

File: MeshLib/Elements/TemplateElement.h [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	/// Returns the ID of a face given an array of nodes.
+	unsigned identifyFace(Node* nodes[3]) const
+	{
+		return ELEMENT_RULE::identifyFace(this->_nodes, nodes);
+	};
+
+	/// Calculates the volume of a convex hexahedron by partitioning it into six tetrahedra.
+	virtual double computeVolume() {return ELEMENT_RULE::computeVolume(this->_nodes);}
```
Comment (norihiro-w) [inline MeshLib/Elements/TemplateElement.h]: could be const?

File: MeshLib/Elements/CellRule.cpp [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	const MathLib::Vector3 c (e->getCenterOfGravity());
+	const unsigned nFaces (e->getNFaces());
+	for (unsigned j=0; j<nFaces; ++j)
+	{
+		MeshLib::Face const*const face (dynamic_cast<const MeshLib::Face*>(e->getFace(j)));
+		const MeshLib::Node x (*(face->getNode(1)));
+		const MathLib::Vector3 cx (c, x);
+		const double s = MathLib::scalarProduct(face->getSurfaceNormal(), cx);
+		delete face;
+		if (s >= 0)
```
Comment (wenqing) [inline MeshLib/Elements/CellRule.cpp]: Compared with real number.

File: MeshLib/Elements/CellRule.h [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+#ifndef CELLRULE_H_
+#define CELLRULE_H_
+
+#include "MeshLib/MeshEnums.h"
+#include "Element.h"
+
+namespace MeshLib {
+
+/**
+ */
```
Comment (wenqing) [inline MeshLib/Elements/CellRule.h]: Documentation?

File: MeshLib/Elements/CellRule.h [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+class CellRule
+{
+public:
+	/// Constant: Dimension of this mesh element
+	static const unsigned dimension = 3u;
```
Comment (wenqing) [inline MeshLib/Elements/CellRule.h]: Only 3D?

File: MeshLib/Elements/Edge.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+	/**
+		* Checks if the node order of an element is correct by testing surface normals.
+		* For 1D elements this always returns true.
+		*/
+	virtual bool testElementNodeOrder() const { return true; }
+
+protected:
+	/// Constructor for a generic mesh element without an array of mesh nodes.
+	Edge(unsigned value = 0, std::size_t id = std::numeric_limits<std::size_t>::max()) : Element(value, id) {}
```
Comment (wenqing) [inline MeshLib/Elements/Edge.h]: Arguments needs doc..

**Comment by norihiro-w** [comment]
This PR is moved to #572

**Pull request closed without merging**